builtin/nomad/jobspec/releaser.go

```diff
@@ -0,0 +1,605 @@
+	return nil
+}
+
+func (r *Releaser) resourceJobDestroy(
+	log hclog.Logger,
+	client *nomadClient,
+	state *Resource_Job,
+	sg terminal.StepGroup,
+) error {
+	log.Trace("No resource destroyed")
+	return nil
+}
+
+func (r *Releaser) resourceJobStatus(
+	ctx context.Context,
+	log hclog.Logger,
+	sg terminal.StepGroup,
+	state *Resource_Job,
+	client *nomadClient,
+	sr *resource.StatusResponse,
+) error {
+	s := sg.Add("Checking status of Nomad job resource %q...", state.Name)
+	defer s.Abort()
+
+	jobClient := client.NomadClient.Jobs()
+	s.Update("Getting job...")
+	// TODO: Because we don't have the namespace from the jobspec, we rely on the
+	//   NOMAD_NAMESPACE env var/searching for job via prefix- consider passing namespace
+	//   from deploy phase
+	jobs, _, err := jobClient.PrefixList(state.Name)
+	if err != nil {
+		return err
+	} else if len(jobs) == 0 {
+		s.Status(terminal.StatusError)
+		s.Update("Job not found.")
+		return nil
+	}
+
+	jobResource := sdk.StatusReport_Resource{
+		CategoryDisplayHint: sdk.ResourceCategoryDisplayHint_INSTANCE_MANAGER,
+	}
+	sr.Resources = append(sr.Resources, &jobResource)
+
+	s.Update("Getting job info...")
+	q := &api.QueryOptions{Namespace: jobs[0].JobSummary.Namespace}
+	job, _, err := jobClient.Info(jobs[0].ID, q)
+
+	if err != nil {
+		s.Update("No job was found")
+		s.Status(terminal.StatusError)
+		s.Done()
+		s = sg.Add("")
+
+		jobResource.Name = state.Name
+		jobResource.Health = sdk.StatusReport_MISSING
+		jobResource.HealthMessage = sdk.StatusReport_MISSING.String()
+	} else if job == nil {
+		s.Update("No job was found")
+		s.Status(terminal.StatusError)
+		s.Done()
+		s = sg.Add("")
+
+		jobResource.Name = state.Name
+		jobResource.Health = sdk.StatusReport_UNKNOWN
+		jobResource.HealthMessage = sdk.StatusReport_UNKNOWN.String()
+	} else {
+		jobResource.Id = *job.ID
+		jobResource.Name = *job.Name
+		jobResource.CreatedTime = timestamppb.New(time.Unix(0, *job.SubmitTime))
+		jobResource.Health = sdk.StatusReport_READY
+		jobResource.HealthMessage = fmt.Sprintf("Job %q exists and is ready", *job.Name)
+		stateJson, err := json.Marshal(map[string]interface{}{
+			"deployment": job,
+		})
+		if err != nil {
+			jobResource.StateJson = string(stateJson)
+		}
+	}
+
+	s.Update("Finished building report for Nomad job resource")
+	s.Done()
+	return nil
+}
+
+// Release promotes the Nomad canary deployment
+func (r *Releaser) Release(
+	ctx context.Context,
+	log hclog.Logger,
+	src *component.Source,
+	job *component.JobInfo,
+	ui terminal.UI,
+	target *Deployment,
+	dcr *component.DeclaredResourcesResp,
+) (*Release, error) {
+	var result Release
+
+	// We'll update the user in real time
+	// TODO: Replace ui.Status with StepGroups once this bug
+	// has been fixed: https://github.com/hashicorp/waypoint/issues/1536
+	st := ui.Status()
+	defer st.Close()
+
+	rm := r.resourceManager(log, dcr)
+	if err := rm.CreateAll(
+		ctx, log, st, &result, target,
```

Comment:
This is a pretty subtle error, but we need to include a `StepGroup` parameter here for `CreateAll`. The reason being, for when a Release encounters a failure, Waypoint automatically "rolls back" to a previous version. We do this through the plugin SDK, and end up calling `DestroyAll` directly rather than the top level `Destroy` here. `DestroyAll` expects a stepgroup, and without it, we get a bad error:

```
! 2 errors occurred:
        * rpc error: code = Aborted desc = Context cancelled from timeout checking
  health of task group "app": context deadline exceeded
        * Error during rollback: 1 error occurred:
        * argument cannot be satisfied: type: terminal.StepGroup. This is a bug in the
  go-argmapper library since this shouldn't happen at this point.
```

This is basically saying when our internal SDK attempted to call `DestroyAll` during a Release failure, the arguments it were given could not satisfy any defined Destroy functions that expected a StepGroup.

For the fix, it is pretty simple:

```diff

diff --git a/builtin/nomad/jobspec/releaser.go b/builtin/nomad/jobspec/releaser.go
index 237218fe8..7cb514b10 100644
--- a/builtin/nomad/jobspec/releaser.go
+++ b/builtin/nomad/jobspec/releaser.go
@@ -319,11 +319,12 @@ func (r *Releaser) Release(
        // TODO: Replace ui.Status with StepGroups once this bug
        // has been fixed: https://github.com/hashicorp/waypoint/issues/1536
        st := ui.Status()
+       sg := ui.StepGroup()
        defer st.Close()
+       defer st.Wait()

        rm := r.resourceManager(log, dcr)
        if err := rm.CreateAll(
-               ctx, log, st, &result, target,
+               ctx, log, st, sg, &result, target,
        ); err != nil {
                return nil, err
        }

```

Reply:
Thanks Brian! Just pushed the update, good catch!